File: MATLAB/utilities/gc1d_setup.m
function NN = gc1d_setup
 
%% network architcture: 2 rings, each N units, one with CW bias and one with CCW bias

    NN.N = 364;

    % Center Surround mexican hat connectivity profile
    A_ext = 1000;
    s_exc = 1.05/100;    
    A_inh = 1000;
    s_inh = 1.00/100;    
    NN = mexHat(NN, A_ext, s_exc, A_inh, s_inh); 

    % synaptic weights (shifts currently hard coded)
    l_RR = -1;          % right to right bias by 1
    l_LL = 1;           % left to left bias by 1
    l_RL=0;
    l_LR=0;
    NN = synapses(NN);

    %global excitatory input
    NN.beta_0 = 100;
    NN.FF_global= NN.beta_0*ones(NN.N,1);    

    % velocity drive parameters
    NN.beta_vel = 1;     %velocity gain

    % topographically organized phase preference    
    NN.x_prefs = (1:NN.N)'/NN.N; 
    
    save NN
end

function NN = mexHat(NN, A_ext, s_exc, A_inh, s_inh)

%% set the profile of the mexican hat connectivity

    z = (-NN.N/2:1:NN.N/2-1);
    NN.mexHat = A_ext*exp(-s_exc*z.^2) - A_inh*exp(-s_inh*z.^2);
    NN.mexHat = circshift(NN.mexHat, [0 NN.N/2 - 1]);

end

function NN = synapses(NN)
    
%% set model synaptic connections

    for i = 1:NN.N
        NN.W_RR(i,:) =  circshift(NN.mexHat ,[0 i-1]); % Right neurons to Right neurons
        NN.W_LL(i,:) =  circshift(NN.mexHat ,[0 i+1]); % Left neurons to Left neurons
        NN.W_RL(i,:) =  circshift(NN.mexHat ,[0 i]);     % Left neurons to Right neurons
        NN.W_LR(i,:) =  circshift(NN.mexHat ,[0 i]);     % Right neurons to Left neurons
    end
    
end

File: Python/utilities/gc1d_setup.py
import numpy as np


class NNStruct:
    def __init__(self):
        pass


def mex_hat(NN, A_ext, s_exc, A_inh, s_inh):
    z = np.arange(-NN.N / 2, NN.N / 2)
    NN.mexHat = A_ext * np.exp(-s_exc * z**2) - A_inh * np.exp(-s_inh * z**2)
    NN.mexHat = np.roll(NN.mexHat, int(NN.N / 2 - 1))
    return NN


def synapses(NN):
    NN.W_RR = np.zeros((NN.N, NN.N))
    NN.W_LL = np.zeros((NN.N, NN.N))
    NN.W_RL = np.zeros((NN.N, NN.N))
    NN.W_LR = np.zeros((NN.N, NN.N))

    for i in range(NN.N):
        NN.W_RR[i, :] = np.roll(NN.mexHat, i)       # Right neurons to Right neurons
        NN.W_LL[i, :] = np.roll(NN.mexHat, i + 2)    # Left neurons to Left neurons
        NN.W_RL[i, :] = np.roll(NN.mexHat, i + 1)    # Left neurons to Right neurons
        NN.W_LR[i, :] = np.roll(NN.mexHat, i + 1)    # Right neurons to Left neurons
    return NN


def gc1d_setup():
    NN = NNStruct()
    NN.N = 364

    # Center Surround mexican hat connectivity profile
    A_ext = 1000
    s_exc = 1.05 / 100
    A_inh = 1000
    s_inh = 1.00 / 100
    NN = mex_hat(NN, A_ext, s_exc, A_inh, s_inh)

    # synaptic weights (shifts currently hard coded)
    NN = synapses(NN)

    # global excitatory input
    NN.beta_0 = 100
    NN.FF_global = NN.beta_0 * np.ones((NN.N, 1))

    # velocity drive parameters
    NN.beta_vel = 1     # velocity gain

    # topographically organized phase preference
    NN.x_prefs = np.arange(1, NN.N + 1) / NN.N

    return NN
